From b34db6a8f81ed3bd7e1065c64ceff1f8322bd6f9 Mon Sep 17 00:00:00 2001
From: Robert Mader <robert.mader@collabora.com>
Date: Wed, 28 Jan 2026 15:46:06 +0100
Subject: [PATCH 6/6] aperture: Prefer openh264 over x264

The former performs significantly better, especially on mobile devices.
While x264 is expected to be the better encoder overall, it's not
totally unexpected that openh264 performs better in the use case at
hand, given it's strong focus on webrtc / calls. Possibly x264 could
archive similar performance with the appropriate options.

This essentially unbreaks video recording on mobile devices such as the
FairPhone 5.

(cherry picked from commit 91dad387e9d5c8521290af111743e9da7714fb40)
---
 aperture/src/viewfinder.rs | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/aperture/src/viewfinder.rs b/aperture/src/viewfinder.rs
index 39cf1c7..90a23b0 100644
--- a/aperture/src/viewfinder.rs
+++ b/aperture/src/viewfinder.rs
@@ -1000,6 +1000,12 @@ impl Viewfinder {
                     }
                     hw_encoder_found = true;
                 }
+                if let Some(encoder) = registry.lookup_feature("openh264enc") {
+                    encoder.set_rank(gst::Rank::PRIMARY);
+                }
+                if let Some(encoder) = registry.lookup_feature("x264enc") {
+                    encoder.set_rank(gst::Rank::PRIMARY - 1);
+                }
                 log::debug!(
                     "Setting up recording with h264/mp4 profile {} hw acceleration",
                     if self.enable_hw_encoding() && hw_encoder_found {
-- 
2.52.0

